p2p/protocols/protocol.go

```diff
@@ -367,6 +388,12 @@ func (p *Peer) Send(ctx context.Context, msg interface{}) error {
 	return p2p.Send(p.rw, code, wmsg)
 }
 
+// EnableMsgPauser enables message pausing for this peer
+// IMPORTANT: to be used only for testing
+func (p *Peer) EnableMsgPauser() {
```

Comment:
Could this method be replaced with `SetMsgPauser(p Pauser)` and to have HandleMsgPauser (instead msgPauserEnabled) on every peer instead globally. Would that be a bit better? This would require syncPauser to be in production code, not only test files. Maybe to mitigate this, to have global pauser interface variable in stream (set to syncPauser only in tests). Maybe that would make protocols api a bit nicer and global pauser unexported in stream.

Reply:
SGTM, let me try and let's see if it looks better!

Reply:
Just pushed the changes, is something like this what you had in mind?